L70-node-proto-loader-type-generator.md

```diff
@@ -0,0 +1,216 @@
+import { ServiceNameHandler } from 'generated/fully/qualified/package/ServiceName';
+
+const packageDefinition = protoLoader.loadSync('other/path/to/proto-file-name.proto', options);
+const loadedPackageDefinition = grpc.loadPackageDefinition(packageDefinition) as unknown as ProtoGrpcType; // How the types will be used
+
+const requestObject: MessageName = {
+  field1: value1,
+  field2: value2
+};
+
+const serviceHandler: ServiceNameHandler = {
+  // Call implicitly has the correct arity (unary, client streaming, etc.) and the call and callback have the correct request and response types
+  methodName(call, callback) {
+    // Implementation
+  }
+}
+
+const server = new grpc.Server();
+server.addService(loadedPackageDefinition.fully.qualified.package.ServiceName.service, serviceHandler);
+
+```
+
+The tool will be named `proto-loader-gen-types` and will have this usage:
+
+```
+proto-loader-gen-types [OPTIONS] <proto file name> ...
+```
+
+The options will correspond directly to the `protoLoader.load` options as follows:
+
+ - `--keepCase`: Preserve the case of field names
+ - `--longs=String|Number`: Specify the type that should be used to output 64 bit integer values
+ - `--enums=String`: Specify that enum values should be output as strings
+ - `--bytes=Array|String`: Specify the type that should be used to output `bytes` fields
+ - `--defaults`: Indicates that default values should be output for missing fields
+ - `--arrays`: Indicates that empty arrays should be output for missing repeated fields even if `--defaults` is unset
+ - `--objects`: Indicates that empty objects should be output for missing message fields even if `--defaults` is unset
+ - `--oneofs`: Indicates that virtual "oneof" fields will be set to the present field's name in the output
+ - `--includeDirs=<directory>`, `-I <directory>`: A directory to search for included `.proto` files. Can be passed multiple times to include multiple directories
+
+ - `--outDir=<directory>`, `-O <directory>`: The directory in which to output files
+ - `--grpcLib=grpc|@grpc/grpc-js`: The gRPC implementation library that these types will be used with
+ - `--verbose`, `-v`: Enable various logging output
+ - `--includeComments`: Include comments from the `.proto` files in the generated files
+
+The output will be one file for each `message` and `enum` loaded, with file paths based on the package and type names, plus a master file per input file that combines all of those to produce the type that the user will load, as described above. `ProtoGrpcType` types from different files can be intersected to get the type that results from loading those files together at runtime. Messages will have an additional type generated, suffixed with `__Output`, that describes the type of objects that will be output by gRPC, i.e. response messages on the client, and request messages on the server. These "output" message types will be subtypes of the main message type, restricted based on the options that are set.

```

Comment:
Could you expand a bit on the need for stricter types, here or under the rationales section?  

Also, I wonder if "_Strict" will work better. Conceivably, I may want to use the strict types for both input and output messages, e.g. to improve the overall type conformity of my codebase.

Reply:
I have reworded the part of the rationales section about the stricter types.